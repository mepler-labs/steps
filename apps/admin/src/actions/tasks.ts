import axios from 'axios';

type DispatchFn = (any) => any;

const apiUrl = process.env.API_URL || 'http://localhost:3001/api';

const CREATE_TASK = 'CREATE_TASK';
export const createTask = async (data): Promise<any> => {
  await axios.post(apiUrl + '/tasks', data);
  return { type: 'foo' };
};

const GET_TASKS = 'GET_TASKS';
export const getTasks = (): DispatchFn => async dispatch => {
  const tasks = await axios.get(apiUrl + '/tasks');
  return dispatch(setTasks(tasks.data));
};

export const SET_TASKS = 'SET_TASKS';
export const setTasks = tasks => {
  return {
    type: SET_TASKS,
    tasks
  };
};

export const SET_TASK_STATUS = 'SET_TASK_STATUS';
export const setTaskStatus = (task, status): DispatchFn => async dispatch => {
  const newTask = { ...task, status };
  delete newTask.steps;

  const tasks = await axios.put(`${apiUrl}/tasks/${task.id}`, newTask);
  return dispatch({ type: SET_TASK_STATUS, id: task.id, status });
}

export const completeTask = (id) => {
  return { type: SET_TASK_STATUS, id, status: 'COMPLETED' };
};

export const setTaskActive = (id, status) => {
  return { type: SET_TASK_STATUS, id, status: 'ACTIVE' };
};

export const setTaskArchived = (id, status) => {
  return { type: SET_TASK_STATUS, id, status: 'ARCHIVED' };
};

export const DELETE_TASK = 'DELETE_TASK';
export const deleteTask = (id): DispatchFn => async dispatch => {
  const tasks = await axios.delete(`${apiUrl}/tasks/${id}`);
  return { type: DELETE_TASK, tasks };
};

export const ADD_TASK = 'ADD_TASK';
export const addTask = (task): DispatchFn => async dispatch => {
  const tasks = await axios.post(apiUrl + '/tasks', task);
  return {
    type: ADD_TASK,
    task
  };
};
