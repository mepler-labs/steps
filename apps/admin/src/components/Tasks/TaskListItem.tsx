import { Box, Flex } from 'grid-styled';
import React from 'react';
import { Link } from 'react-router-dom';
import styled from 'styled-components';
import { blue, green, lightGrey, white } from 'styles/colors';
import { remCalc, sansSerif } from 'styles/type';

const StyledLink = styled(Link)`
  color: ${green};
  font-size: 0.8em;
  margin-top: 0.5em;
  text-decoration: none;
  text-transform: uppercase;
`;

export default ({ setTaskStatus, task, url, index }) => {
  const toggleTaskStatus = e => {
    const status = task.status === 'COMPLETED' ? 'ACTIVE' : 'COMPLETED';
    setTaskStatus(task, status);
  };

  const checked =
    task.status === 'COMPLETED' ? 'check_circle' : 'check_circle_outline';

  return (
    <Background
      flexDirection="column"
      alignItems="flex-start"
      justifyContent="center"
      p={20}
      width={1}
    >
      <Flex flexDirection="row" alignItems="center">
        <div onClick={toggleTaskStatus}>
          <i className={`material-icons ${checked}`}>{checked}</i>
        </div>
        <Title>{task.title}</Title>
      </Flex>
      <Box ml={34}>
        <StyledLink to={`${url}/${task.id}`}>View Steps &rarr;</StyledLink>
      </Box>
    </Background>
  );
};

const Background = styled(Flex)`
  background: ${white};
  border-bottom-right-radius: 4px;
  border-top-right-radius: 4px;

  i.material-icons {
    cursor: pointer;
    width: 24px;
  }

  .check_circle {
    color: ${blue};
  }

  .check_circle_outline {
    color: ${lightGrey};
  }
`;

const Title = styled.h3`
  font-family: ${sansSerif};
  font-size: ${remCalc(24)};
  font-weight: 500;
  margin-top: 0;
  margin-left: 10px;
  margin-right: 0;
  margin-bottom: ${remCalc(10)};
`;
