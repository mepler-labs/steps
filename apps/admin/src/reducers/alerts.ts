export default (state = [], action) => {
  switch (action.type) {
    case 'ADD_ALERT':
      return [
        ...state,
        {
          message: action.message,
          level: action.level,
          id: action.id
        }
      ];

    case 'REMOVE_ALERT':
      return state.filter(alert => {
        if (alert.id === action.id) {
          return false;
        } else {
          return true;
        }
      });

    default:
      return state;
  }
};
