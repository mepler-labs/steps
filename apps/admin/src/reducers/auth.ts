import { LOGIN, LOGOUT, SET_USER_TYPE } from 'actions/auth';

export enum USER_TYPE {
  SUPER_ADMIN = 'Superadmin',
  ADMIN = 'Admin',
  COACH = 'Coach',
  CLIENT = 'Client',
}

export type Org = {
  id?: number;
  name: string;
  sms_number: string;
  logo?: string;
};

export type User = {
  checkin_times?: any;
  coach_id?: number;
  color?: string;
  email?: string;
  first_name?: string;
  follow_up_date?: any;
  goals?: any;
  id?: number;
  image?: any;
  last_name?: string;
  org_id?: number;
  phone?: string;
  plan_url?: string;
  platform?: string;
  status?: string;
  type: null | USER_TYPE;
  updated?: string;
  org?: Org;
};

export interface State {
  user: null | User;
  isAuthenticated: boolean;
}

const Storage = {
  get: key => {
    const data = localStorage.getItem(key);
    return JSON.parse(data);
  },

  set: (key, value) => {
    localStorage.setItem(key, JSON.stringify(value));
  },

  remove: key => {
    localStorage.removeItem(key);
  },
};

const initialState: State = {
  user: Storage.get('USER') || { type: null },
  isAuthenticated: Storage.get('AUTHENTICATED') || false,
};

export default (state = initialState, action): State => {
  switch (action.type) {
    case SET_USER_TYPE:
      return { ...state, user: { type: action.userType } };

    case LOGIN:
      Storage.set('USER', action.user);
      Storage.set('AUTHENTICATED', true);
      return { ...state, user: action.user, isAuthenticated: true };

    case LOGOUT:
      Storage.remove('USER');
      Storage.remove('AUTHENTICATED');
      return { ...state, user: { type: null }, isAuthenticated: false };

    default:
      return state;
  }
};
