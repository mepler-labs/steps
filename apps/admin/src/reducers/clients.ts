import { UPDATE_CLIENT } from './../actions/clients';
import {
  ADD_CLIENT,
  SET_CLIENTS,
  SET_CLIENT_FOLLOW_UP_DATE,
  SET_CLIENT_GOALS,
  SET_CLIENT_MESSAGES,
  SET_CLIENT_REQUESTS,
} from 'actions/clients';

export enum USER_PLATFORM {
  SMS = 'SMS',
  FACEBOOK = 'FBOOK',
}

export type UserPlatform = USER_PLATFORM.SMS | USER_PLATFORM.FACEBOOK;
export type UserStatus = 'AWAITING_HELP' | 'WORKING' | 'NON_RESPONSIVE';

export type Client = {
  id: number;
  first_name: string;
  last_name: string;
  email: string;
  phone: string;
  coach_id: number;
  org_id: number;
  color: string;
  goals: string[];
  status: UserStatus;
  updated: string;
  platform: UserPlatform;
  image: string;
  follow_up_date: string;
  checkin_times: CheckinTime[];
  topic: string;
};

export type CheckinTime = {
  topic: string;
  message: string;
  time: string;
};

export type Org = {
  id: number;
  name: string;
  sms_number: string;
  logo: string;
};

export interface ClientsState {
  clients: Client[];
  clientMessages: any;
  clientRequests: any;
}

const initialState: ClientsState = {
  clients: [],
  clientMessages: {},
  clientRequests: {},
};

export default (state = initialState, action) => {
  switch (action.type) {
    case SET_CLIENTS: {
      return {
        ...state,
        clients: action.clients,
      };
    }
    case ADD_CLIENT: {
      return {
        ...state,
        clients: [...state.clients, action.client],
      };
    }
    case SET_CLIENT_GOALS: {
      return {
        ...state,
        clients: state.clients.map(client => {
          if (client.id === action.clientId) {
            return {
              ...client,
              goals: action.goals,
            };
          }
          return client;
        }),
      };
    }
    case SET_CLIENT_MESSAGES: {
      return {
        ...state,
        clientMessages: Object.assign(state.clientMessages, {
          [action.clientId]: action.messages,
        }),
      };
    }
    case SET_CLIENT_REQUESTS: {
      return {
        ...state,
        clientRequests: Object.assign(state.clientRequests, {
          [action.clientId]: action.requests,
        }),
      };
    }
    case SET_CLIENT_FOLLOW_UP_DATE: {
      return {
        ...state,
        clients: state.clients.map(client => {
          if (client.id === action.clientId) {
            return {
              ...client,
              follow_up_date: action.followUpDate,
            };
          }
          return client;
        }),
      };
    }
    case UPDATE_CLIENT: {
      return {
        ...state,
        clients: state.clients.map(client => {
          if (client.id === action.client.id) {
            return action.client;
          }
          return client;
        }),
      };
    }
    default: {
      return state;
    }
  }
};
