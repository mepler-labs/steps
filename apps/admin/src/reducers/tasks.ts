import { SET_TASKS, ADD_TASK, SET_TASK_STATUS } from 'actions/tasks';

export type TaskStatus = 'ACTIVE' | 'COMPLETED' | 'ARCHIVED';
export type TaskDifficulty = 'EASY' | 'MODERATE' | 'DIFFICULT';

export type Task = {
  id: number;
  title: string;
  category: string;
  description: string;
  status: TaskStatus;
  created_by: number;
  user_id: number;
  difficulty: TaskDifficulty;
  date_created: string;
  date_completed: string;
  recurring: {};
  steps: Step[];
};

export type Step = {
  text: string;
  note: string;
  task_id: number;
};

export interface TasksState {
  tasks: Task[];
}

const initialState: TasksState = {
  tasks: [
    {
      id: 0,
      title: 'string',
      category: 'string',
      description: 'string',
      status: 'ACTIVE',
      created_by: 0,
      user_id: 0,
      difficulty: 'EASY',
      date_created: '2018-06-04T14:44:14.901Z',
      date_completed: '2018-06-04T14:44:14.901Z',
      recurring: {},
      steps: [
        {
          text: 'string',
          note: 'string',
          task_id: 0,
        },
      ],
    },
  ],
};

export default (state = initialState, action) => {
  if (action.type === SET_TASKS) {
    return {
      ...state,
      tasks: action.tasks,
    };
  }

  if (action.type === ADD_TASK) {
    return {
      ...state,
      tasks: [...state.tasks, action.task],
    };
  }

  if (action.type === SET_TASK_STATUS) {
    return {
      ...state,
      tasks: state.tasks.map(task => {
        if (task.id !== action.id) return task;
        return { ...task, status: action.status };
      }),
    };
  }

  return state;
};
