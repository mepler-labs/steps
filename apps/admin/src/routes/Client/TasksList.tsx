import { setTaskStatus, getTasks } from 'actions/tasks';
import TaskList from 'components/Tasks/TaskList';
import { filterById } from 'helpers';
import React from 'react';
import { connect } from 'react-redux';
import { Match } from 'react-router-dom';
import { Task } from 'reducers/tasks';
import { bindActionCreators } from 'redux';
import { AlertLevel } from 'components/Alert/types';
import { addAlert } from 'actions/alerts';
import TaskListItem from 'components/Tasks/TaskListItem';

interface Props {
  tasks: Task[];
  match: Match;
  actions: { setTaskStatus; getTasks; addAlert };
}

class Tasks extends React.Component<Props> {
  componentWillMount() {
    const { actions } = this.props;
    actions.getTasks().catch(err => {
      actions.addAlert({
        level: AlertLevel.Error,
        message: err.message,
        id: 'client-task-list-error',
      });
    });
  }

  render() {
    const { tasks, match, actions } = this.props;
    return (
      <TaskList
        items={tasks}
        setTaskStatus={actions.setTaskStatus}
        url={match.url}
      >
        {childProps => <TaskListItem {...childProps} />}
      </TaskList>
    );
  }
}

const mapStateToProps = state => ({
  tasks: filterById(state.tasks, state.auth.user.id, 'user_id'),
  user: state.auth.user,
});

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators({ setTaskStatus, getTasks, addAlert }, dispatch),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Tasks);
