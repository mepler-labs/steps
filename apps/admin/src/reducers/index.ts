import { combineReducers } from 'redux';
import { reducer as formReducer } from 'redux-form';
import { reducer as storageMergeReducer } from 'redux-storage';
import clients from './clients';
import alerts from './alerts';
import user from './user';
import tasks from './tasks';

export default storageMergeReducer(
  combineReducers({
    form: formReducer,
    clients,
    tasks,
    user,
    alerts
  })
);
