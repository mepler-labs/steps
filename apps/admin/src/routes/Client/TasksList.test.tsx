import 'jest';
import React from 'react';
import { Tasks } from './TasksList';
import { shallow } from 'enzyme';
import { Task } from 'reducers/tasks';
import TaskListItem from 'components/Tasks/TaskListItem';

describe('TasksList.tsx', () => {
  it('gets tasks', () => {
    const actions = {
      getTasks: jest.fn().mockReturnValue(Promise.resolve()),
      setTaskStatus: jest.fn(),
      addAlert: jest.fn(),
    };
    shallow(<Tasks actions={actions} tasks={[]} match={{ url: 'foo/bar' }} />);

    expect(actions.getTasks).toHaveBeenCalled();
  });

  it('displays task list items', () => {
    const actions = {
      getTasks: jest.fn().mockReturnValue(Promise.resolve()),
      setTaskStatus: jest.fn(),
      addAlert: jest.fn(),
    };
    const tasks: Task[] = [
      {
        id: 1,
        title: 'My first task',
      },
      {
        id: 1,
        title: 'My second task',
      },
    ];
    const wrapper = shallow(
      <Tasks actions={actions} tasks={tasks} match={{ url: 'foo/bar' }} />,
    );

    const taskListItems = wrapper.dive().find(TaskListItem);

    expect(wrapper).toMatchSnapshot();
    expect(taskListItems).toHaveLength(2);
  });
});
