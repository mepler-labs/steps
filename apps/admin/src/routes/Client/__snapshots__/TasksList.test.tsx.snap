// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TasksList.tsx displays task list items 1`] = `
<TaskList
  items={
    Array [
      Object {
        "id": 1,
        "title": "My first task",
      },
      Object {
        "id": 1,
        "title": "My second task",
      },
    ]
  }
  setTaskStatus={[MockFunction]}
  url="foo/bar"
/>
`;
